File: Raven.Client.Lightweight/Document/Batches/LazyQueryOperation.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Text;
using Raven.Abstractions.Data;
using Raven.Client.Connection;
using Raven.Client.Document.SessionOperations;
using Raven.Client.Shard;
using Raven.Json.Linq;
using System.Linq;

namespace Raven.Client.Document.Batches
{
    public class LazyQueryOperation<T> : ILazyOperation
    {
        private readonly QueryOperation queryOperation;
        private readonly Action<QueryResult> afterQueryExecuted;
        private readonly HashSet<string> includes;

        private NameValueCollection headers;

        public LazyQueryOperation(QueryOperation queryOperation, Action<QueryResult> afterQueryExecuted, HashSet<string> includes, NameValueCollection headers)
        {
            this.queryOperation = queryOperation;
            this.afterQueryExecuted = afterQueryExecuted;
            this.includes = includes;
            this.headers = headers;
        }

        public GetRequest CreateRequest()
        {
            var stringBuilder = new StringBuilder();
            queryOperation.IndexQuery.AppendQueryString(stringBuilder);

            foreach (var include in includes)
            {
                stringBuilder.Append("&include=").Append(include);
            }
            var request = new GetRequest
            {
                Url = "/indexes/" + queryOperation.IndexName, 
                Query = stringBuilder.ToString()
            };
            if (headers != null)
            {
                foreach (var headerKey in headers.Keys)
                {
                    request.Headers[headerKey.ToString()] = headers[headerKey.ToString()];
                }
            }
            return request;
        }

        public object Result { get; set; }

        public QueryResult QueryResult { get; set; }

        public bool RequiresRetry { get; set; }
        public void HandleResponses(GetResponse[] responses, ShardStrategy shardStrategy)
        {
            var count = responses.Count(x => x.Status == 404);
            if (count != 0)
            {
                throw new InvalidOperationException("There is no index named: " + queryOperation.IndexName + " in " + count + " shards");
            }

            var list = responses
                .Select(response => SerializationHelper.ToQueryResult((RavenJObject)response.Result, response.Headers["Temp-Request-Time"], -1))
                .ToList();

            var queryResult = shardStrategy.MergeQueryResults(queryOperation.IndexQuery, list);

            RequiresRetry = queryOperation.IsAcceptable(queryResult) == false;
            if (RequiresRetry)
                return;

            if (afterQueryExecuted != null)
                afterQueryExecuted(queryResult);
            Result = queryOperation.Complete<T>();
            QueryResult = queryResult;
        }

        public void HandleResponse(GetResponse response)
        {
            if (response.ForceRetry)
            {
                Result = null;
                RequiresRetry = true;
                return;
            }

            if (response.Status == 404)
                throw new InvalidOperationException("There is no index named: " + queryOperation.IndexName + Environment.NewLine + response.Result);
            var json = (RavenJObject)response.Result;
            var queryResult = SerializationHelper.ToQueryResult(json, response.Headers["Temp-Request-Time"], -1);
            HandleResponse(queryResult);
        }

        private void HandleResponse(QueryResult queryResult)
        {
            RequiresRetry = queryOperation.IsAcceptable(queryResult) == false;
            if (RequiresRetry)
                return;

            if (afterQueryExecuted != null)
                afterQueryExecuted(queryResult);
            Result = queryOperation.Complete<T>();
            QueryResult = queryResult;
        }

        public IDisposable EnterContext()
        {
            return queryOperation.EnterQueryContext();
        }

        public void SetHeaders(NameValueCollection theHeaders)
        {
            headers = theHeaders;
        }
    }
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/document/batches/LazyQueryOperation.java
package net.ravendb.client.document.batches;

import java.util.Map;
import java.util.Set;

import net.ravendb.abstractions.closure.Action1;
import net.ravendb.abstractions.data.GetRequest;
import net.ravendb.abstractions.data.GetResponse;
import net.ravendb.abstractions.data.QueryResult;
import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.client.connection.HttpExtensions;
import net.ravendb.client.connection.SerializationHelper;
import net.ravendb.client.document.sessionoperations.QueryOperation;

import org.apache.http.HttpStatus;


public class LazyQueryOperation<T> implements ILazyOperation {
  private final QueryOperation queryOperation;
  private final Action1<QueryResult> afterQueryExecuted;
  private final Set<String> includes;
  private final Class<T> clazz;

  private QueryResult queryResult;

  private Map<String, String> headers;

  private Object result;
  private boolean requiresRetry;

  public LazyQueryOperation(Class<T> clazz, QueryOperation queryOperation, Action1<QueryResult> afterQueryExecuted, Set<String> includes) {
    this.clazz = clazz;
    this.queryOperation = queryOperation;
    this.afterQueryExecuted = afterQueryExecuted;
    this.includes = includes;
  }

  @Override
  public QueryResult getQueryResult() {
    return queryResult;
  }

  public void setQueryResult(QueryResult queryResult) {
    this.queryResult = queryResult;
  }



  @Override
  public GetRequest createRequest() {
    StringBuilder stringBuilder = new StringBuilder();
    queryOperation.getIndexQuery().appendQueryString(stringBuilder);

    for (String include : includes) {
      stringBuilder.append("&include=").append(include);
    }
    GetRequest request = new GetRequest();
    request.setUrl("/indexes/" + queryOperation.getIndexName());
    request.setQuery(stringBuilder.toString());

    if (headers != null) {
      for (Map.Entry<String, String> header : headers.entrySet()) {
        request.getHeaders().put(header.getKey(), header.getValue());
      }
    }
    return request;
  }

  @Override
  public Object getResult() {
    return result;
  }

  public void setResult(Object result) {
    this.result = result;
  }

  @Override
  public boolean isRequiresRetry() {
    return requiresRetry;
  }

  public void setRequiresRetry(boolean requiresRetry) {
    this.requiresRetry = requiresRetry;
  }

  @Override
  public void handleResponse(GetResponse response) {
    if (response.getStatus() == HttpStatus.SC_NOT_FOUND) {
      throw new IllegalStateException("There is no index named: " + queryOperation.getIndexName() + "\n" + response.getResult());
    }
    RavenJObject json = (RavenJObject)response.getResult();
    QueryResult queryResult = SerializationHelper.toQueryResult(json, HttpExtensions.getEtagHeader(response), response.getHeaders().get("Temp-Request-Time"));
    handleResponse(queryResult);
  }

  private void handleResponse(QueryResult queryResult) {
    requiresRetry = queryOperation.isAcceptable(queryResult) == false;
    if (requiresRetry) {
      return;
    }

    if (afterQueryExecuted != null)
      afterQueryExecuted.apply(queryResult);
    result = queryOperation.complete(clazz);
    this.queryResult = queryResult;
  }

  @Override
  public AutoCloseable enterContext() {
    return queryOperation.enterQueryContext();
  }

  public void setHeaders(Map<String, String> headers) {
    this.headers = headers;
  }


}
